metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,285 @@
+  private case class InvalidQuery(uri: URI)
+      extends Error(s"Invalid query `${uri.getQuery}`")
+  private case class InvalidQueryParams(
+      params: Array[Array[String]],
+      override val ex: Throwable
+  ) extends ErrorWithException(
+        s"Invalid parameters `${params.map(_.mkString("=")).mkString(",")}`",
+        ex
+      )
+  private case class MissingDecoder(uri: URI)
+      extends Error(s"Missing `decoder` argument `$uri`")
+  private case class MissingOriginalURI(uri: URI)
+      extends Error(s"Missing `originalURI` argument `$uri`")
+  private case class InvalidURIPath(path: String, override val ex: Throwable)
+      extends ErrorWithException(s"Invalid URI path `$path`", ex)
+  private case class FileMissing(path: AbsolutePath)
+      extends Error(s"File missing `$path`")
+  private case class InvalidDecoder(decoder: String)
+      extends Error(s"Invalid `$decoder`")
+  private case class FileToDecodeMissing(path: AbsolutePath)
+      extends Error(s"""|No file to decode for $path
+                        |Is Metals still starting?
+                        |Is the file compiled?
+                        |""".stripMargin)
+  private case class DecoderError(
+      path: AbsolutePath,
+      override val ex: Throwable
+  ) extends ErrorWithException(s"Error decoding file `$path`", ex)
+
+  private case class Decoder(
+      findFile: AbsolutePath => Option[AbsolutePath],
+      decode: AbsolutePath => Future[Either[Error, String]]
+  )
+
+  /**
+   * URI format...
+   * metalsDecodeFile:/fileUrl?decoder=decoderName&keyValueArguments&originalURI=originalFileURI
+   *
+   * Examples...
+   * metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap&originalURI=file:///somedir/someFile.java
+   * metalsDecodeFile:/somedir/someFile.class.javap.verbose?decoder=javap&originalURI=file:///somedir/someFile.class
+   * metalsDecodeFile:/somedir/someFile.scala?decoder=semanticdb&originalURI=file:///somedir/someFile.scala
+   * metalsDecodeFile:/somedir/someFile.scala.semanticdb?decoder=semanticdb&originalURI=file:///somedir/someFile.scala.semanticdb
+   *
+   * errors get logged and sent back as the text contents
+   */
+  def decodedFileContents(uriAsStr: String): Future[String] = {
+    for {
+      check <- Future { getDecodeInfo(uriAsStr) }
+      output <- check match {
+        case Left(value) => Future.successful(Left(value))
+        case Right((decoder, path)) => decoder.decode(path)
+      }
+      checkedOutput = output match {
+        case Left(e: ErrorWithException) =>
+          scribe.error(e.message, e.ex)
+          val sw = new StringWriter
+          e.ex.printStackTrace(new PrintWriter(sw))
+          s"""|${e.message}
+              |${sw.toString}""".stripMargin
+        case Left(e) =>
+          scribe.error(e.message)
+          e.message
+        case Right(success) => success
+      }
+    } yield checkedOutput
+  }
+
+  private def getDecodeInfo(
```

Comment:
Could we simplify this a bit and could we instead of having a separate error for each, just show one error at the end if anything is amiss.

It could show something like:
```
Unexpected uri, Metals accepts ones such as:
metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap&originalURI=file:///somedir/someFile.java
Take a look at scala/meta/internal/metals/FileDecoderProvider for more information.
```
This will only show up for people creating the plugins, so that should be alright.

Alternatively, we can link to the section about the command in https://scalameta.org/metals/docs/integrations/new-editor

We  could simplify it a bit in the for comprehension, added some examples for that.